lib/rules/no-reject-literal.js

```diff
@@ -0,0 +1,125 @@
+/**
+ * @fileoverview restrict values that can be used as Promise rejection reasons
+ * @author Teddy Katz
+ */
+"use strict";
+
+const astUtils = require("../ast-utils");
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+module.exports = {
+    meta: {
+        docs: {
+            description: "disallow using literals as Promise rejection reasons",
+            category: "Best Practices",
+            recommended: false
+        },
+        fixable: null,
+        schema: []
+    },
+
+    create(context) {
+
+        const resolveRejectFuncs = new WeakSet();
+        const rejectFuncIdentifiers = new WeakSet();
+
+        //----------------------------------------------------------------------
+        // Helpers
+        //----------------------------------------------------------------------
+
+        /**
+        * Checks the argument of a reject() or Promise.reject() CallExpression, and reports it if it can't be an Error
+        * @param {ASTNode} callExpression A CallExpression node which is used to reject a Promise
+        * @returns {void}
+        */
+        function checkRejectedLiteral(callExpression) {
+            if (
+                !callExpression.arguments.length ||
+                !astUtils.couldBeError(callExpression.arguments[0]) ||
+                callExpression.arguments[0].type === "Identifier" && callExpression.arguments[0].name === "undefined"
+            ) {
+                context.report({
+                    node: callExpression,
+                    message: "Expected the Promise rejection reason to be an Error."
+                });
+            }
+        }
+
+        /**
+        * Checks references to the `reject` argument in a resolveRejectFunc, which is a function(resolve, reject) { }) that was passed
+        * as an argument to the Promise constructor.
+        * @param {ASTNode} node The `FunctionExpression` or `ArrowFunctionExpression` node that was passed
+        * as an argument to the Promise constructor. This function must have at least two parameters, and the second parameter
+        * must be an `Identifier` (i.e. no destructuring). Also, this must be the node that is currently being traversed.
+        * @returns {void}
+        */
+        function checkResolveRejectFunc(node) {
+            if (resolveRejectFuncs.has(node)) {
+
+                // Get all variables in the function's scope.
+                context.getScope().variables
```

Comment:
I think we can use `context.getDeclaredVariables(node)` instead of `context.getScope().variables` to simplify the finding logic (I guess `WeakSet`s are unnecessary).
If the `node` is a function, `context.getDeclaredVariables(node)` returns an array which includes the variables of the function name and parameters.